### PR TITLE
Decommit instance memory after a runtime call on Linux

### PR DESCRIPTION
**Pull request by pepyakin**
Right now we have a wasm runtime cache. This cache keeps a number of instantiated runtimes around. That allows the executor to serve the requests with as little latency as possible. This improves latency on the critical paths significantly.

However, the instances retain their memory between executions. This was made as an optimization. This means that if a runtime wrote into some page some physical memory will stay attached to it.

In the light of increasing the heap pages (https://github.com/paritytech/substrate/pull/8892), this problem becomes more severe. With the defaults, that can lead to up 2 GiB consumed by those dirty mappings.

This PR solves this problem on Linux. The solution is to tell the kernel that we are no longer interested in the contents of the specified address range. This will lead to demouting these pages and freeing up the corresponding areas of physical memory. This will not lead to decreasing of virtual memory usage by the instances.

As a side effect on Linux this brings us back in line with https://github.com/paritytech/substrate/pull/3011 after we dropped it. There is no intention to bring this back as a guarantee as of yet. In other words, wasm runtimes must not access uninitialized memory.

### CODE REVIEW

File: client/executor/src/integration_tests/mod.rs [hunk 1]
```diff
@@ -775,3 +784,76 @@ fn panic_in_spawned_instance_panics_on_joining_its_result(wasm_method: WasmExecu
...
+			let rss = line.parse::<u32>()
+				.expect("failed to parse the RSS");
+
+			return rss
+		}
+
+		panic!("smaps doesn't contain rss!");
+	}
+
+	// We test this only under unix because
```
Comment (shawntabrizi) [inline client/executor/src/integration_tests/mod.rs:834]: is there more to the comment?

Comment (pepyakin) [reply client/executor/src/integration_tests/mod.rs]: Nah, it's no longer relevant.

File: client/executor/src/integration_tests/mod.rs [hunk 2]
```diff
@@ -775,3 +784,82 @@ fn panic_in_spawned_instance_panics_on_joining_its_result(wasm_method: WasmExecu
...
+		// However, in practice the memory usage can either go up and sometimes down (hence
+		// `saturating_sub` below). This is because this test is run along with other concurrently
+		// and the other tests may allocate or deallocate memory. This in theory makes this test
+		// flaky. This problem should be excaberated with the number of CPU cores
+		//
+		// However, empirically, this works just fine on the "build host" (numcpus=128) and is able
+		// reliably to tell if the decommiting really works or not.
+		//
+		// That said, if the test fails randomly we should consider turning it off.
+		assert!(rss_post.saturating_sub(rss_pre) < 32768 /* kB */);
```
Comment (koute) [inline client/executor/src/integration_tests/mod.rs:863]: Wouldn't it be better to just look at the RSS of the specific pages where the WASM heap is kept instead of the whole process?

E.g. something like this test program:

```rust
use std::ops::Range;
use std::collections::BTreeMap;

fn get_smaps() -> Vec<(Range<usize>, BTreeMap<String, usize>)> {
    let regex_start = regex::RegexBuilder::new("^([0-9a-f]+)-([0-9a-f]+)").multi_line(true).build().unwrap();
    let regex_kv = regex::RegexBuilder::new(r#"^([^:]+):\s*(\d+) kB"#).multi_line(true).build().unwrap();
    let smaps = std::fs::read_to_string("/proc/self/smaps").unwrap();
    let boundaries: Vec<_> = regex_start
        .find_iter(&smaps)
        .map(|matched| matched.start())
        .chain(std::iter::once(smaps.len()))
        .collect();

    let mut output = Vec::new();
    for window in boundaries.windows(2) {
        let chunk = &smaps[window[0]..window[1]];
        let caps = regex_start.captures(chunk).unwrap();
        let start = usize::from_str_radix(caps.get(1).unwrap().as_str(), 16).unwrap();
        let end = usize::from_str_radix(caps.get(2).unwrap().as_str(), 16).unwrap();

        let values = regex_kv.captures_iter(chunk).map(|cap| {
            let key = cap.get(1).unwrap().as_str().to_owned();
            let value = cap.get(2).unwrap().as_str().parse().unwrap();
            (key, value)
        }).collect();

        output.push((start..end, values));
    }
                                                                                                                                      
    output                                                                                                                            
}                                                                                                                                     
                                                                                                                                      
fn get_map(addr: usize) -> BTreeMap<String, usize> {                                                                                  
    get_smaps().into_iter().find(|(range, _)| {                                                                                       
        addr >= range.start && addr < range.end                                                                                       
    }).unwrap().1                                                                                                                       
}

fn get_rss(addr: usize) -> usize {
    *get_map(addr).get("Rss").unwrap()
}

fn main() {
    unsafe {
        let p = libc::mmap(
            std::ptr::null_mut(),
            4096 * 16,
            libc::PROT_READ|libc::PROT_WRITE,
            libc::MAP_ANONYMOUS|libc::MAP_PRIVATE,
            0,
            0
        );

        let pp = p as *mut u8;
        println!("Initial RSS:");
        println!("{}", get_rss(pp as usize));

        println!("After dirtying one page:");
        pp.write(1);
        println!("{}", get_rss(pp as usize));

        println!("After dirtying two pages:");
        pp.add(4096).write(1);
        println!("{}", get_rss(pp as usize));

        println!("After MADV_DONTNEED:");
        libc::madvise(p, 4096 * 16, libc::MADV_DONTNEED);
        println!("{}", get_rss(pp as usize));
    }
}
```

Output:

```
Initial RSS:
0
After dirtying one page:
4
After dirtying two pages:
8
After MADV_DONTNEED:
0
```

Comment (pepyakin) [reply client/executor/src/integration_tests/mod.rs:863]: Yeah, good idea, that would be definitely more precise! It would require somehow plumbing the base addr of the linear memory here though. That sounds burdensome and not sure if it worth the effort right now IMO. OTOH, if this test proves to be flaky then that would be the way to go.

WDYT?

Comment (koute) [reply client/executor/src/integration_tests/mod.rs:863]: Well, it's up to you. (:

Personally I think it's worthwhile *if* we want to have a test for this, since with the way the test is currently written its long-term value is somewhat questionable - you could both have it fail when it should succeed, and succeed when it should fail, so effectively does it *actually* protect us from a regression when it's not guaranteed to give a correct result?

But again, it's up to you. You know this code a lot better than I do after all!

Comment (pepyakin) [reply client/executor/src/integration_tests/mod.rs]: You convinced me.

See https://github.com/paritytech/substrate/pull/8998/commits/5d07a203269dab07a9404e7e5aecf08cdab40587 and later PRs.

The trait fn is annoying but it is not that much hassle.

Comment (koute) [reply client/executor/src/integration_tests/mod.rs]: Looks good to me!

File: client/executor/src/integration_tests/linux.rs [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+	instance
+		.call_export(
+			"test_dirty_plenty_memory",
+			&(heap_base as u32, 1024u32).encode(),
+		)
+		.unwrap();
+	let rss_post = smaps.get_rss(base_addr).unwrap();
+
+	assert_eq!(rss_post.saturating_sub(rss_pre), 0);
```
[the remark this answers is not in the document]
Comment (pepyakin) [reply client/executor/src/integration_tests/linux.rs:67]: Ah, but that'll regress the error message !

[the remark this answers is not in the document]
Comment (athei) [reply client/executor/src/integration_tests/linux.rs:67]: I don't like making the expression for complicated in order to save writing a manual error message. Just pass that you think is appropriate as a third argument.

That said, this is of course a minor nit and you can leave it this way .

File: client/executor/src/integration_tests/mod.rs [hunk 4]
```diff
@@ -64,6 +67,15 @@ macro_rules! test_wasm_execution {
...
 	};
+
+	(compiled_only $method_name:ident) => {
+		paste::item! {
+			#[test]
+			fn [<$method_name _compiled>]() {
+				$method_name(WasmExecutionMethod::Compiled);
+			}
+		}
+	};
```
Comment (athei) [inline client/executor/src/integration_tests/mod.rs:78]: I don't see this used anywhere. Why is this added here?

File: client/executor/src/integration_tests/linux.rs [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+	let base_addr = instance.linear_memory_base_ptr().unwrap() as usize;
+	let rss_pre = smaps.get_rss(base_addr).unwrap();
+
+	instance
+		.call_export(
+			"test_dirty_plenty_memory",
+			&(heap_base as u32, 1024u32).encode(),
+		)
+		.unwrap();
+	let rss_post = smaps.get_rss(base_addr).unwrap();
```
Comment (athei) [inline client/executor/src/integration_tests/linux.rs:65]: As far as I see it the information is not refetched from the `proc` here. I recommend converting this into a one shot `get_rss` function that fetches when you call it.

File: client/executor/src/integration_tests/linux.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (athei) [reply client/executor/src/integration_tests/linux.rs:67]: Shouldn't it test for equality because all memory is decommissioned?

File: client/executor/src/integration_tests/linux.rs [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+use crate::WasmExecutionMethod;
+use super::mk_test_runtime;
+use codec::Encode as _;
+
+mod smaps;
+
+use self::smaps::Smaps;
+
+#[test]
+fn memory_consumption_compiled() {
```
Comment (athei) [inline client/executor/src/integration_tests/linux.rs:35]: Please add some text what this test is checking. I am not sure whether it could confirm that the memory was dirtied or that it was correctly decommissioned.

**Review by athei — CHANGES_REQUESTED** [review]
I believe the test is incorrect.

File: client/executor/src/integration_tests/mod.rs (hunk 4 above)
Comment (pepyakin) [reply client/executor/src/integration_tests/mod.rs:78]: Good catch. It was used in previous iteration but not anymore

File: client/executor/runtime-test/src/lib.rs [hunk 7]
```diff
@@ -69,6 +69,36 @@ sp_core::wasm_export_functions! {
...
+		let heap_ptr = heap_ptr as *mut u8;
+
+		// Traverse the host pages and make each one dirty
+		let host_pages = heap_pages as usize * 16;
+		for i in 0..host_pages {
+			unsafe {
+				// technically this is an UB, but there is no way Rust can find this out.
+				heap_ptr.add(i * 4096).write(0);
+			}
+		}
```
Comment (athei) [inline client/executor/runtime-test/src/lib.rs:95]: Apple Silicon has 16KiB page size. Note that you can also run Linux on Apple Silicon (soon).

Comment (pepyakin) [reply client/executor/runtime-test/src/lib.rs:95]: Yes, I know. But here it should be fine since we will hit such a page several times. I do not think this should be an issue, or?

Comment (athei) [reply client/executor/runtime-test/src/lib.rs:95]: Yeah you are right. For this code it would only be a problem if  would be run on an architecture with a smaller page size. I don't know any that could run substrate.

**Comment by pepyakin** [comment]
Thanks for the review, that was a great catch! It should be fine now.

File: client/executor/src/integration_tests/linux/smaps.rs [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+		&self.0
+			.iter()
+			.find(|(range, _)| addr >= range.start && addr < range.end)
+			.unwrap()
+			.1
+	}
+
+	pub fn get_rss(&self, addr: usize) -> Option<usize> {
+		self.get_map(addr).get("Rss").cloned()
+	}
```
Comment (athei) [inline client/executor/src/integration_tests/linux/smaps.rs:81]: How sure are you that the whole instance memory is kept in one mapping? I think they can also be split up in multiple consecutive mappings (at least on macOS that is the case).

Comment (pepyakin) [reply client/executor/src/integration_tests/linux/smaps.rs:81]: This is Linux only code so we don't care about macOS here. Yes, potentially they can be split up on several mappings. The opposite is also possible AFAIU, when several instances are packed into a single VMA. Note though that the solution with `madvise` should work in these cases.

We generally cannot be 100% sure what wasmtime does under the hood, so this is the best effort. But hey it's not the end of the world if this test fails. I see no reason it to be not a single mapping though, at least on Linux. It is simpler to implement, most efficient, etc

File: client/executor/wasmtime/src/instance_wrapper.rs [hunk 9]
```diff
@@ -415,6 +415,43 @@ impl InstanceWrapper {
...
+
+	/// Removes physical backing from the allocated linear memory. This leads to returning the memory
+	/// back to the system. While the memory is zeroed this is considered as a side-effect and is not
+	/// relied upon. Thus this function acts as a hint.
+	pub fn decommit(&self) {
+		if self.memory.data_size() == 0 {
+			return;
+		}
+
+		cfg_if::cfg_if! {
```
Comment (bkchr) [inline client/executor/wasmtime/src/instance_wrapper.rs:432]: We don't need `cfg_if` here?

**Review by bkchr — APPROVED** [review]

File: client/executor/wasmtime/src/instance_wrapper.rs (hunk 9 above)
Comment (pepyakin) [reply client/executor/wasmtime/src/instance_wrapper.rs:432]: You mean that `madvise` is POSIX-compatible? Sure, but it's complicated. In Linux this is not merely an advice it has a very certain and widely relied semantics of eagerly freeing and zeroing the pages. macOS is, on the other hand, is very lazy about those. madvise(WONTNEED) won't have the same effect there.

There are different options there to make it work on macOS, but those are out of scope for this PR. When the time comes though it will be clear where to integrate them : )

Comment (bkchr) [reply client/executor/wasmtime/src/instance_wrapper.rs:432]: I mean we don't need the `cfg_if`.

Could be just `cfg(unix)`

Comment (pepyakin) [reply client/executor/wasmtime/src/instance_wrapper.rs:432]: (Assuming you meant `cfg(target_os = "linux")`)

Yeah, true, not necessary. I figured I use it because:

- it's not a new dependency. We already use it and it is a very common dep that it occurs in our dep tree many times
- as I mentioned it's likely we want other branches for macOS and/or `cfg(unix)`

It's not something I am married to, so can change to a plain `cfg` as well. Just let me know

Comment (bkchr) [reply client/executor/wasmtime/src/instance_wrapper.rs:432]: Yeah fine

**Comment by bkchr** [comment]
bot merge
